crate/operator/restore_backup.py

```diff
@@ -359,6 +365,140 @@ async def shards_recovery_in_progress(
             raise kopf.PermanentError("Shards could not be fetched.")
 
 
+class RestoreType(abc.ABC):
+    """
+    Base class for the different types of restore operations. New subclasses
+    are identified by their ``restore_type`` and can be added dynamically by
+    using the ``register_subclass`` decorator.
+    """
+
+    subclasses: Dict[str, Any] = {}
+
+    def __init__(
+        self,
+        *,
+        tables: Optional[List[str]] = None,
+        sections: Optional[List[str]] = None,
+        partitions: Optional[List[Dict]] = None,
+        **_kwargs,
+    ):
+        self.tables = tables
+        self.sections = sections
+        self.partitions = partitions
+
+    @classmethod
+    def register_subclass(cls, restore_type: str):
+        def decorator(subclass):
+            cls.subclasses[restore_type] = subclass
+            return subclass
+
+        return decorator
+
+    @classmethod
+    def create(cls, restore_type: str, *args, **kwargs):
+        if restore_type not in cls.subclasses:
+            raise kopf.PermanentError(f"Unknown restore type {restore_type}")
+
+        return cls.subclasses[restore_type](*args, **kwargs)
+
+    @abc.abstractmethod
+    def get_restore_keyword(self):
+        """
+        Each subclass needs to return the keyword to be used in the
+        ``RESTORE SNAPSHOT`` command based on the type of restore operation.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    async def validate_restore_complete(
+        self, *, cursor: Cursor, snapshot: str, logger: logging.Logger, **_kwargs
+    ):
+        """
+        Each subclass needs to define a method which indicates that the restore
+        operation has been completed successfully. Or raise a ``kopf.TemporaryError``
+        in case the operation is still in progress.
+        """
+        raise NotImplementedError()
+
+
+@RestoreType.register_subclass(RESTORE_TYPE_TABLES)
+class RestoreTables(RestoreType):
+    def get_restore_keyword(self):
+        tables = self.tables or []
+        # keep this check for backwards compatibility
+        if not tables or (len(tables) == 1 and tables[0].lower() == "all"):
+            return "ALL"
+
+        return f'TABLE {",".join(tables)}'
+
+    async def validate_restore_complete(
+        self, *, cursor: Cursor, snapshot: str, logger: logging.Logger, **_kwargs
+    ):
+        tables = self.tables or []
+        await shards_recovery_in_progress(cursor, snapshot, tables, logger)
+
+
+@RestoreType.register_subclass(RESTORE_TYPE_METADATA)
+class RestoreMetadata(RestoreType):
+    def get_restore_keyword(self):
+        return "METADATA"
+
+    async def validate_restore_complete(
+        self, *, cursor: Cursor, snapshot: str, logger: logging.Logger, **_kwargs
+    ):
+        return True
```

Comment:
Should this method really invariably return `True`? Would that still be the case if the operation failed for any reason?

Reply:
`validate_restore_complete()` only runs after the actual restore operation is finished, e.g. to check if all tables have been restored successfully for operations which include table-restores. For metadata I do not really know how to validate if the metadata has been restored successfully, so in this case it is considered successful as soon as the check for `sys.snapshot_restore` table [here](https://github.com/crate/crate-operator/blob/master/crate/operator/restore_backup.py#L195) succeeds.

If the restore operation itself fails, it still reports the error already before.